ocaml/src/verify.ml

```diff
@@ -0,0 +1,166 @@
+(*
+Copyright 2015 Open vStorage NV
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*)
+
+open Prelude
+open Slice
+open Lwt_bytes2
+open Lwt.Infix
+
+let verify_and_maybe_repair_object
+      (alba_client : Alba_base_client.client)
+      ~namespace_id
+      ~verify_checksum
+      ~repair_osd_unavailable
+      manifest =
+  let open Nsm_model in
+  let open Manifest in
+  let object_id = manifest.object_id in
+
+  let osd_access = alba_client # osd_access in
+
+  Lwt_list.mapi_s
+    (fun chunk_id chunk_location ->
+     Lwt_list.mapi_p
+       (fun fragment_id location ->
+        match location with
+        | (None, _) -> Lwt.return `NoneOsd
+        | (Some osd_id, version_id) ->
+
+           let key_string =
+             Osd_keys.AlbaInstance.fragment
+               ~namespace_id
+               ~object_id ~version_id
+               ~chunk_id ~fragment_id
+           in
+           let key = Slice.wrap_string key_string in
+
+           Lwt.catch
+             (fun () ->
+              osd_access # with_osd
+                ~osd_id
+                (fun osd ->
+                 if verify_checksum
```

Comment:
This is constant for as far as the maps are concerned; it's probably better to build a function
verify: osd_id -> key (* list *) -> [< `ChecksumMismatch | `Missing | `Ok ]
up front, that you pass down.